Diffs and Dry runs, Take II.

dgou opened the pull request:
Shares all but the last two commits with #2 
I think I like this a lot better,
but will wait for @peterkaminski to weigh in before updating version and README.

dgou committed:
Use Path earlier, leverage .samefile

Slightly more robust, but mostly for simplifying
converting from strings to Path(s).

dgou committed:
Deduplicate copy file and directory helpers code.

dgou committed:
Use new copy helper and delete old ones.

Did this as two commits because the git diff was horrible otherwise.

dgou committed:
Leverage new copy helper by using a list of items to be copied.

dgou committed:
Add --verbose option.

Typically -v is the short-cut for verbose and -V for --version,
(see python --help for example), but -v was already in use for --version.

So I decided not to add a short-hand for --verbose yet,
pending a decision from Peter on what direction to take.

dgou committed:
Make operational flags mutually exclusive.

This leaves `--rm` in a weird position,
since it doesn't have any effect unless --update is used.

dgou committed:
Refactor into action helpers in prep for dry-run option.

Also use subprocess.run, it's one less step.

dgou committed:
Add --dry-run option

dgou committed:
Move execute into it's own helper too.

dgou committed:
Simplify backup code with a unified remove helper.

Doing this in two steps because git diff mess.
And another simplification as well coming up.

dgou committed:
Remove unused helpers.

Part 2 of 2 to keep diffs simpler.

dgou committed:
Move date generating and matching info close together.

Reasons:
1. Easier to keep in sync if in the unlikely event it changes.
2. Keeping it together is easier to conceptualize.
3. No need to keep assigning the glob each time a function changes
   since it doesn't depend on any function parameters or time
   (just time format)

dgou committed:
Minor cleanup to how execute works.

1. Move print into the helper to simplify call from main
   and prepare for additional simplifications to come.
2. For helpers that take the vault path, always put it first
   to keep it easier to remember and to make simplifications easier.

dgou committed:
Start dedup-ing looping code to simplify main function.

More simplifications to come, but this is a nice starting chunk.

dgou committed:
Use new all-vault helper for backup operations.

Also, try to keep knowledge of the command line args contained
to the main function to make future changes easier.

dgou committed:
Use all vault helper for listing vaults.

Small simplification, but will also be handy in next commit too.

dgou committed:
Check that the update vault is really one of the user's vaults.

dgou committed:
copy vault: Minor change to enable additional simplifications.

Reasons:
1. Having the destination as the first parameter is more like an assignment statement.
2. That's what C's memcpy does and it's a pretty common pattern.
3. Fits the pattern needed for the all vault iteration helper.

dgou committed:
Use all vault helper for copying vaults.

Small change but easier to track parameter changes vs. doing this
in the previous commit

dgou committed:
Cleanup cleanup helper, keeps args in the main function.

dgou committed:
Fix spine-o: Use single quotes instead of double-suotes

My day-job convention is to use double-quotes, so when I am typing
my spine picks the quote characters without any conscience effort.
This brings all the changes back into compliance with the code base
as it was before they were added.

dgou committed:
Fix minor coding style issues.

dgou committed:
Minor cleanups, doc strings, redundant comments, etc.

dgou committed:
Spike out a potential solution for doing diff.

Next commit will implement this.

dgou committed:
Fix subprocess output vs. python buffering.

As per the previous commit comment, python buffering can cause
print output to come out after subprocess command output.
There are number of ways to fix that (StackOverflow has so many
and many duplicates), but since we're using /usr/bin/env anyways,
the -S option lets you specify additional command line parameters
to the invoked program, so this feels like the simplest solution.
Assuming it works on a Mac for Peter of course!

dgou committed:
Trying a simpler diff implementation.

While it does duplicate copy_settings a bit,
I think this is overall cleaner than the first PR.

dgou commented on osm.py:
```diff
@@ -172,20 +180,19 @@ def copy_settings(src, dest, args):
     dir_path = Path(vault_path) / '.obsidian'
-    iso_8601_glob = '*-[0-9][0-9][0-9][0-9]-[0-9][0-9]-[0-9][0-9]T[0-9][0-9]:[0-9][0-9]:[0-9][0-9].[0-9]*Z'
-    for dest in dir_path.glob(iso_8601_glob):
+    for dest in dir_path.glob(ISO_8601_GLOB):
```
To clarify the typo in the commit message above,
the value of the glob doesn't depend on any of the parameters here, and it doesn't depend on something gobal changing at run time (such as time or other state), so it's clearer to indicate that by using a global.

dgou commented on osm.py:
```diff
@@ -243,7 +246,7 @@ def main():
         elif args.update:
             ensure_valid_vault(vault_paths, args.update)
-            call_for_each_vault(copy_settings, Path.home() / args.update, args)
+            call_for_each_vault(vault_paths, copy_settings, Path.home() / args.update, args.rm)
```
Miseed this in the previous commit!

dgou commented on osm.py:
```diff
@@ -243,8 +243,7 @@ def main():
             ensure_valid_vault(vault_paths, args.update)
-            for vault_path in vault_paths:
-                copy_settings(vault_path, Path.home() / args.update, args)
+            call_for_each_vault(copy_settings, Path.home() / args.update, args)
```
Next commit fixes the missing vault_paths parameter needed here.

dgou commented on osm.py:
```diff
@@ -248,6 +262,10 @@ def main():
             call_for_each_vault(vault_paths, copy_settings, Path.home() / args.update, args.rm)
+        elif args.diff_to:
+            ensure_valid_vault(vault_paths, args.diff_to)
+            # Note: By setting DIFF_CMD above, we can re-use the copy_settings helper.
```
This will change in the next commit, but it was easier to leave it here in this commit because I based this branch off the end of #2

dgou commented on osm.py:
```diff
@@ -223,6 +220,34 @@ def copy_settings(dest, src, clean_first):
+        src_item = src / item
+        if not src_item.exists():
+            continue
+        diff_cmd = [DIFF_CMD, "-r", dest_item, src_item]
```
Note that on Linux `-r` works even for files.
If this doesn't work on a Mac then we can make it conditional.
Also note that by using the command list form here, the extra shell process isn't needed and more importantly we avoid having to quote paths with spaces, or other funny shell characters.
For the execute command we do want the shell involved because it is a user typed command, not one predefined like this one.

peterkaminski replied:
On my Mac, the `-r` makes it so it only diffs dirs (and subdirs), not files. So no file diff is ever output afaict.

Maybe related, I have a number of vaults, and they have different sets of directories in them. I get a bunch of things on stderr, e.g.:

```
... /snippets: No such file or directory
... /plugins: No such file or directory
```

etc.

Because of the buffering stuff, all these errors show up in one bunch in the middle of all the stdout.

So, I guess my wish is:

- files get diffed, as well as dirs/subdirs, if that was the intent you were going for (I could play around with different diffs, maybe I just need to install something from homebrew)
- that stderr output gets suppressed (i think?)

dgou replied:
The buffering was fixed in a later commit, and I see you fixed my typos there, thank you!
As to the `-r` I don't have a mac any more so I don't know the installed diff,
but making it only use `-r` on directories is fine w/me.
Also, I will fix the `No such fie or directory` and report better when one side of the diff doesn't exist,
my own test cases locally didn't catch that, thanks!

dgou replied:
I think the diff issues have been fixed, but wordsmithing the new messages is welcome!

peterkaminski committed:
fix two tiny typos

peterkaminski committed:
small output change

peterkaminski commented:
It's looking really good! Thanks for doing all this. :-)

Along with another comment above about `diff -r`, a couple of observations:

A) In commit https://github.com/peterkaminski/obsidian-settings-manager/pull/3/commits/9e0566a0e144f3905991de21a5b8ec860d571c65, I added `\n# ` to the print statement to make the output a little nicer, similar to another print elsewhere. I had chosen `#` because it's sort of a comment in a text file (at least for unix-ish people) and also a header for markdown people.

B) I purposely didn't have my settings "template" vault open in Obsidian (it's empty, and just used to host settings), so I got this:

```
./osm.py -n -u /Users/peterkaminski/obsidian-settings
Error: '/Users/peterkaminski/obsidian-settings' is not one of your vaults:
(list of all vaults)
```

I think it's probably reasonable just to open the "obsidian-settings" vault in Obsidian, and then it works fine. So unless you think differently, I wouldn't change anything for this.

C) One of my wishlist items in `## Possible Enhancements` in the README is "Output commands to run on the OS's CLI, rather than actually executing actions", and the DRY-RUN output is tantalizingly close to being able to be redirected into a text file and then sh'ed.  Wdyt about reformatting the DRY-RUN output so it would be easy to run the output as a shell script, or adding another flag command that outputs something without the DRY-RUN comments that could be run as a shell script?

dgou committed:
Refactor diff into a helper and only call when both sides exist.

Next commit will report when one side doesn't so user knows what
an update will do.

dgou commented:
A) Pulled your commits, so I got those changes, thanks!
B) I am not following on this part, I don't have that issue?  Or, I think when you had the TODO in the code that implmented, there wasn't any mention of a vault settings directory not part of the Obsidian config itself?
C) Yeah, the dry run is really close.  I didn't want to go too far in this PR until I got your feedback.
A few new commits about to post for the diff changes and once that's OKd I will update the readme/version
This is as far as I wanted to go to get to the diff output.
The dry-run commands will be a little trickier because of how the DRY_RUN printing is done now, so I don't have a quick follow up for that. Once this merges I will think about it, but no promises, whichever of us gets there first! :-)
Thank you for being so open to these changes, that's really cool and I appreciate it!

peterkaminski commented:
Thanks for all the work, I super appreciate it!

For C), no rush! If/whenever it works out, and yeah, whichever of us gets there first. :-)

Regarding B), don't worry about it.

The slightly longer answer (feel free to ignore): I didn't even look, but I think the way the helpers work, they expect that the source vault is one of the vaults Obsidian knows about. I specifically didn't tell Obsidian about my settings source vault, so OSM errored out. The workaround/fix is just to tell Obsidian about that vault.

The addendum to the slightly longer answer: I'm at 50+ vaults at this point, some of them real, and some for various kinds of testing this and that, and it made sense to have sort of an "out-of-band" vault to host all the settings, without ever loading it into Obsidian. Most people probably won't have that much confusion with all their vaults. :-)

dgou committed:
Don't run diff unless both parts exist, better reporting when they don't

dgou committed:
Update version and README, make this a real PR.

dgou commented:
B)  Ah, I kinda surmised that as I was working on the diff commits.
I think having "really, don't check if this is a used vault, I know what I am doing" option would be easy.
I did make the major options exclusive, but some of the options which affect other options don't fit well in that model.
argparse is powerful, but has some limitations. Some of the limitations are that I am only at level 6 of 10 in using it :-)

C) I would like to do that, but am out of time today to keep going on code. I will probably wait until I want it more or until I have an AHA. And maybe you'll have gotten there first by then. :-)

peterkaminski committed:
document new stuff in README, bump (c) date

dgou commented on README.md:
```diff
@@ -97,6 +113,16 @@ The files and directories currently copied (as long as they exist in the source
 - `snippets` - CSS snippets for customizing Obsidian's appearance
 - `README.md` - optional file used for vaults distributed to others via Git
 
+See the `ITEMS_TO_COPY` global in the code to configure this.
```
Yeah, I wish this were easier to keep in sync.
What I have done in other work projects is tell folks to run the program with a `--show-items` command, which is like `--list` here, but for ITEMS_TO_COPY instead of the vault list.
I prefer self-documenting code as much as possible, and self-describing code, but I see the value of having this written down here for folks kicking the tires.

dgou replied:
Of course that is moot if the list is lifted out of the code. Just thinking out loud.

dgou commented on README.md:
```diff
@@ -142,13 +168,13 @@ Bug reports, enhancement suggestions, and pull requests are welcome at the [Obsi
 
 - List details of vault plugins, snippets, config.
-- More fine-grained control over which files and directories are copied, and which destination vaults they are copied to.
+- More fine-grained control over which files and directories are copied, and which destination vaults they are copied to. This would replace the `ITEMS_TO_COPY` global.
```
Thinking about this, probably have a config file somewhere to get that list completely out of the code.

peterkaminski committed:
small typo fix

peterkaminski committed:
microseconds are always 6 digits, zero-padded

peterkaminski committed:
add backup-create, backup-restore to wishlist

peterkaminski commented:
I think it's good, but I'm going on break right now, so I'll merge later today.

And lol, I was trying out the update/backup-remove stuff, and edited my root obsidian.json file so it was only one vault affected instead of all; I managed to trash my backup copy of obsidian.json, so I need to regenerate it, with all 50+ vaults.

dgou commented on osm.py:
```diff
@@ -38,7 +38,7 @@ def datestring():
 
 # Keep this in sync with the format returned by datestring()
-ISO_8601_GLOB = '*-[0-9][0-9][0-9][0-9]-[0-9][0-9]-[0-9][0-9]T[0-9][0-9]:[0-9][0-9]:[0-9][0-9].[0-9]*Z'
+ISO_8601_GLOB = '*-[0-9][0-9][0-9][0-9]-[0-9][0-9]-[0-9][0-9]T[0-9][0-9]:[0-9][0-9]:[0-9][0-9].[0-9][0-9][0-9][0-9][0-9][0-9]Z'
```
It's really too bad globs don't support the regex count syntax `{4}` and such.
But then they'd be regexes :-)
I am almost ready to add a new commit that does:
`YEAR = '[0-9]' * 4` and such, but... :-)

peterkaminski replied:
I feel ya, lol.

I actually _swapped in_ all the `{}` sections, then realized it was a glob and not a regex, so I reverted.

peterkaminski committed:
minor wording changes

peterkaminski committed:
minor wording changes

peterkaminski commented:
All set, merging!
